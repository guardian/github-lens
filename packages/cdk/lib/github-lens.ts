import { GuApiLambda, GuScheduledLambda } from '@guardian/cdk';
import { GuCertificate } from '@guardian/cdk/lib/constructs/acm';
import type { NoMonitoring } from '@guardian/cdk/lib/constructs/cloudwatch';
import { GuStack, GuStringParameter } from '@guardian/cdk/lib/constructs/core';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuCname } from '@guardian/cdk/lib/constructs/dns';
import { GuS3Bucket } from '@guardian/cdk/lib/constructs/s3';
import type { App } from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import { Schedule } from 'aws-cdk-lib/aws-events';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Key } from 'aws-cdk-lib/aws-kms';
import { Runtime } from 'aws-cdk-lib/aws-lambda';

export class GithubLens extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps, domainName: string) {
		super(scope, id, props);

		const app = 'github-lens';

		const dataBucket = new GuS3Bucket(this, `${app}-data-bucket`, {
			app,
		});

		const kmsKeyAlias = `${this.stage}/${this.stack}/${app}`;
		const kmsKey = new Key(this, kmsKeyAlias, {
			enableKeyRotation: true,
		});

		const kmsDecryptPolicy = new PolicyStatement({
			effect: Effect.ALLOW,
			actions: ['kms:Decrypt'],
			resources: [kmsKey.keyArn],
		});

		const paramPathBase = `/${this.stage}/${this.stack}/${app}`;
		const repoFetcherApp = 'repo-fetcher';
		const apiApp = 'github-lens-api';

		const githubAppId = new GuStringParameter(this, 'github-app-id', {
			default: `${paramPathBase}/github-app-id`,
			description:
				'(From SSM) The GitHub app ID of the app used to authenticate github-lens',
			fromSSM: true,
		});

		const githubInstallationId = new GuStringParameter(
			this,
			'github-installation-id',
			{
				default: `${paramPathBase}/github-installation-id`,
				description:
					'(From SSM) The GitHub installation ID of the app used to authenticate github-lens in the Guardian org',
				fromSSM: true,
			},
		);

		const githubPrivateKey = new GuStringParameter(this, 'github-private-key', {
			default: `${paramPathBase}/github-private-key`,
			noEcho: true,
			description:
				'(From SSM) (KMS encrypted) The private key of the app used to authenticate github-lens in the Guardian org',
			fromSSM: true,
		});

		const noMonitoring: NoMonitoring = { noMonitoring: true };

		const apiLambda = new GuApiLambda(this, `${apiApp}-lambda`, {
			fileName: `${apiApp}.zip`,
			handler: 'handler.main',
			runtime: Runtime.NODEJS_16_X,
			monitoringConfiguration: noMonitoring,
			app: apiApp,
			api: {
				id: 'github-lens',
				description: 'API that proxies all requests to Lambda',
			},
		});

		const domain = apiLambda.api.addDomainName('domain', {
			domainName: domainName,
			certificate: new GuCertificate(this, {
				app: app,
				domainName: domainName,
			}),
		});

		new GuCname(this, 'DNS', {
			app: app,
			ttl: Duration.days(1),
			domainName: domainName,
			resourceRecord: domain.domainNameAliasDomainName,
		});

		const scheduledLambda = new GuScheduledLambda(
			this,
			`${repoFetcherApp}-lambda`,
			{
				app: repoFetcherApp,
				runtime: Runtime.NODEJS_16_X,
				memorySize: 512,
				handler: 'handler.main',
				fileName: `${repoFetcherApp}.zip`,
				monitoringConfiguration: {
					toleratedErrorPercentage: 0,
					snsTopicName: 'devx-alerts',
				},
				rules: [{ schedule: Schedule.cron({ minute: '0', hour: '8' }) }],
				timeout: Duration.seconds(300),
				environment: {
					STAGE: this.stage,
					KMS_KEY_ID: kmsKey.keyId,
					GITHUB_APP_ID: githubAppId.valueAsString,
					GITHUB_APP_PRIVATE_KEY: githubPrivateKey.valueAsString,
					GITHUB_APP_INSTALLATION_ID: githubInstallationId.valueAsString,
					DATA_BUCKET_NAME: dataBucket.bucketName,
				},
				initialPolicy: [kmsDecryptPolicy],
			},
		);

		dataBucket.grantRead(apiLambda);
		dataBucket.grantReadWrite(scheduledLambda);
	}
}
