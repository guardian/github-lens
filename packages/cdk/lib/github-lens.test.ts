import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { GithubLens } from './github-lens';

describe('The GithubLens stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new GithubLens(
			app,
			'GithubLens',
			{
				stack: 'deploy',
				stage: 'INFRA',
			},
			'github-lens.gutools.co.uk',
		);
		const template = Template.fromStack(stack);
		expect(template.toJSON()).toMatchSnapshot();
	});
});
