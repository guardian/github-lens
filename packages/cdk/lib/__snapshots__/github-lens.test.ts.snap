// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GithubLens stack matches the snapshot 1`] = `
Object {
  "Metadata": Object {
    "gu:cdk:constructs": Array [
      "GuS3Bucket",
      "GuStringParameter",
      "GuStringParameter",
      "GuStringParameter",
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
      "GuLambdaErrorPercentageAlarm",
    ],
    "gu:cdk:version": "TEST",
  },
  "Outputs": Object {
    "lensapiEndpoint62764F13": Object {
      "Value": Object {
        "Fn::Join": Array [
          "",
          Array [
            "https://",
            Object {
              "Ref": "lensapi6D87001F",
            },
            ".execute-api.",
            Object {
              "Ref": "AWS::Region",
            },
            ".",
            Object {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            Object {
              "Ref": "lensapiDeploymentStageprodDBC7B676",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "githubappid": Object {
      "Default": "/TEST/deploy/github-lens/github-app-id",
      "Description": "(From SSM) The GitHub app ID of the app used to authenticate github-lens",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "githubinstallationid": Object {
      "Default": "/TEST/deploy/github-lens/github-installation-id",
      "Description": "(From SSM) The GitHub installation ID of the app used to authenticate github-lens in the Guardian org",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "githubprivatekey": Object {
      "Default": "/TEST/deploy/github-lens/github-private-key",
      "Description": "(From SSM) (KMS encrypted) The private key of the app used to authenticate github-lens in the Guardian org",
      "NoEcho": true,
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TESTdeploygithublensFF348C57": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "EnableKeyRotation": true,
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/github-lens",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "githublensdatabucketGithublens7C07FDE7": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "github-lens",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/github-lens",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "lensapi6D87001F": Object {
      "Properties": Object {
        "Body": Object {
          "components": Object {
            "x-amazon-apigateway-integrations": Object {
              "mocked": Object {
                "type": "mock",
              },
            },
          },
          "info": Object {
            "description": "Returns= information about The Guardian's various GitHub projects.",
            "title": "GitHub Metadata API",
            "version": "0.0.1",
          },
          "openapi": "3.0.0",
          "paths": Object {
            "/repos": Object {
              "get": Object {
                "description": "Optional extended description in CommonMark or HTML.",
                "parameters": Array [
                  Object {
                    "description": "Maximum number of entries to return",
                    "in": "path",
                    "name": "limit",
                    "schema": Object {
                      "default": 100,
                      "format": "int64",
                      "maximum": 100,
                      "minimum": 1,
                      "type": "integer",
                    },
                  },
                  Object {
                    "description": "Pagination mechanism.",
                    "in": "path",
                    "name": "offset",
                    "schema": Object {
                      "default": 0,
                      "format": "int64",
                      "minimum": 0,
                      "type": "integer",
                    },
                  },
                ],
                "responses": Object {
                  "200": Object {
                    "content": Object {
                      "application/json": Object {
                        "schema": Object {
                          "items": Object {
                            "properties": Object {
                              "default_branch": Object {
                                "description": "Typically master or main.",
                                "type": "string",
                              },
                              "name": Object {
                                "description": "Short name of the repo",
                                "type": "string",
                              },
                              "owners": Object {
                                "description": "Which GitHub teams are designated as repository 'Owners'",
                                "items": Object {
                                  "type": "string",
                                },
                                "type": "array",
                              },
                              "topics": Object {
                                "description": "How has the repository been tagged? Used to determine production status.",
                                "items": Object {
                                  "type": "string",
                                },
                                "type": "array",
                              },
                            },
                            "type": "object",
                          },
                          "type": "array",
                        },
                      },
                    },
                    "description": "A JSON array of user names",
                  },
                  "404": Object {
                    "content": Object {
                      "application/json": Object {
                        "schema": Object {
                          "type": "string",
                        },
                      },
                    },
                    "description": "An error message, if the user requests data that doesn't exist (ie the offset is too large)",
                  },
                  "500": Object {
                    "content": Object {
                      "application/json": Object {
                        "schema": Object {
                          "type": "string",
                        },
                      },
                    },
                    "description": "Internal server error",
                  },
                },
                "summary": "Returns a list of repo objects, representing un-archived repositories in the guardian organisation.",
                "x-amazon-apigateway-integration": Object {
                  "$ref": "#/components/x-amazon-apigateway-integrations/mocked",
                },
              },
            },
          },
          "servers": Array [
            Object {
              "description": "Optional server description, e.g. Main (production) server",
              "url": "http://repos.gutools.co.uk/v1",
            },
            Object {
              "description": "Optional server description, e.g. Internal staging server for testing",
              "url": "http://repos.test.gutools.co.uk/v1",
            },
          ],
        },
        "Name": "lens-api",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/github-lens",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "lensapiAccount6B4C6C4E": Object {
      "DeletionPolicy": "Retain",
      "DependsOn": Array [
        "lensapi6D87001F",
      ],
      "Properties": Object {
        "CloudWatchRoleArn": Object {
          "Fn::GetAtt": Array [
            "lensapiCloudWatchRole67910D5E",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "lensapiCloudWatchRole67910D5E": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/github-lens",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "lensapiDeploymentA1B5CA24dd1313c773e74549f225347d00b5446d": Object {
      "Properties": Object {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": Object {
          "Ref": "lensapi6D87001F",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "lensapiDeploymentStageprodDBC7B676": Object {
      "Properties": Object {
        "DeploymentId": Object {
          "Ref": "lensapiDeploymentA1B5CA24dd1313c773e74549f225347d00b5446d",
        },
        "RestApiId": Object {
          "Ref": "lensapi6D87001F",
        },
        "StageName": "prod",
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/github-lens",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "repofetcherlambda3F53E41B": Object {
      "DependsOn": Array [
        "repofetcherlambdaServiceRoleDefaultPolicy50E545F9",
        "repofetcherlambdaServiceRole28143926",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "deploy/TEST/repo-fetcher/repo-fetcher.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "APP": "repo-fetcher",
            "DATA_BUCKET_NAME": Object {
              "Ref": "githublensdatabucketGithublens7C07FDE7",
            },
            "GITHUB_APP_ID": Object {
              "Ref": "githubappid",
            },
            "GITHUB_APP_INSTALLATION_ID": Object {
              "Ref": "githubinstallationid",
            },
            "GITHUB_APP_PRIVATE_KEY": Object {
              "Ref": "githubprivatekey",
            },
            "KMS_KEY_ID": Object {
              "Ref": "TESTdeploygithublensFF348C57",
            },
            "STACK": "deploy",
            "STAGE": "TEST",
          },
        },
        "Handler": "handler.main",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "repofetcherlambdaServiceRole28143926",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "repo-fetcher",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/github-lens",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "repofetcherlambdaErrorPercentageAlarmForLambda25B92F1F": Object {
      "Properties": Object {
        "ActionsEnabled": true,
        "AlarmActions": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:aws:sns:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":",
                Object {
                  "Ref": "AWS::AccountId",
                },
                ":devx-alerts",
              ],
            ],
          },
        ],
        "AlarmDescription": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "repofetcherlambda3F53E41B",
              },
              " exceeded 0% error rate",
            ],
          ],
        },
        "AlarmName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "High error % from ",
              Object {
                "Ref": "repofetcherlambda3F53E41B",
              },
              " lambda in TEST",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": Array [
          Object {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "Error % of ",
                  Object {
                    "Ref": "repofetcherlambda3F53E41B",
                  },
                ],
              ],
            },
          },
          Object {
            "Id": "m1",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "FunctionName",
                    "Value": Object {
                      "Ref": "repofetcherlambda3F53E41B",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          Object {
            "Id": "m2",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "FunctionName",
                    "Value": Object {
                      "Ref": "repofetcherlambda3F53E41B",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "repofetcherlambdaServiceRole28143926": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "repo-fetcher",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/github-lens",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": "TEST",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "repofetcherlambdaServiceRoleDefaultPolicy50E545F9": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:PutObject",
                "s3:PutObjectAcl",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "githublensdatabucketGithublens7C07FDE7",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": "kms:Decrypt",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "TESTdeploygithublensFF348C57",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                      "/deploy/TEST/repo-fetcher/repo-fetcher.zip",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/deploy/repo-fetcher",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/TEST/deploy/repo-fetcher/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "repofetcherlambdaServiceRoleDefaultPolicy50E545F9",
        "Roles": Array [
          Object {
            "Ref": "repofetcherlambdaServiceRole28143926",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "repofetcherlambdarepofetcherlambdacron080AllowEventRuleGithubLensrepofetcherlambda584DB7587F4F78A1": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "repofetcherlambda3F53E41B",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "repofetcherlambdarepofetcherlambdacron080B8E1215D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "repofetcherlambdarepofetcherlambdacron080B8E1215D": Object {
      "Properties": Object {
        "ScheduleExpression": "cron(0 8 * * ? *)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "repofetcherlambda3F53E41B",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
}
`;
